core/src/cell.rs

```diff
@@ -69,14 +72,92 @@ impl CellStatus {
     pub fn is_unknown(&self) -> bool {
         self == &CellStatus::Unknown
     }
+
+    pub fn is_unspecified(&self) -> bool {
+        self == &CellStatus::Unspecified
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum HeaderStatus {
+    /// Header exists on current chain
+    Live(Box<Header>),
+    /// Header exists, but the specified block doesn't contain referenced transaction.
+    InclusionFaliure,
+    /// Header does not exist on current chain
+    Unknown,
+    /// OutPoint doesn't contain reference to a header.
+    Unspecified,
+}
+
+impl HeaderStatus {
+    pub fn live_header(header: Header) -> HeaderStatus {
+        HeaderStatus::Live(Box::new(header))
+    }
+
+    pub fn is_live(&self) -> bool {
+        match *self {
+            HeaderStatus::Live(_) => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_inclusion_failure(&self) -> bool {
+        self == &HeaderStatus::InclusionFaliure
+    }
+
+    pub fn is_unknown(&self) -> bool {
+        self == &HeaderStatus::Unknown
+    }
+
+    pub fn is_unspecified(&self) -> bool {
+        self == &HeaderStatus::Unspecified
+    }
+}
+
+#[derive(Debug)]
+pub struct ResolvedOutPoint {
```

Comment:
Both `Header` and `CellMeta` are big in memory, especially `Header`. Considering most `OutPoint` will resolve to cell only, an `Option<Header>` just wastes abundant memory. We need to consider whether using `Option<Box<Header>>` here.